# Исследование и анализ базы данных книжного сервиса

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. 

Наша первая задача как аналитика — проанализировать базу данных.

В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Задания:**

- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc">
<a href='#1'>1. Изучение данных</a>

<a href='#1.1'>1.1 Изучение Таблицы books</a>

<a href='#1.2'>1.2 Изучение Таблицы authors</a>

<a href='#1.3'>1.3 Изучение Таблицы publishers</a>

<a href='#1.4'>1.4 Изучение Таблицы ratings</a>

<a href='#1.5'>1.5 Изучение Таблицы reviews</a>

<a href='#2'>2. Задания</a>
    
<a href='#2.1'>2.1 Сколько книг вышло после 1 января 2000 года</a>   
    
<a href='#2.2'>2.2 Для каждой книги посчитать количество обзоров и среднюю оценку</a>   
    
<a href='#2.3'>2.3 Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры</a>   
    
<a href='#2.4'>2.4 Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками</a>   
    
<a href='#2.5'>2.5 среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a>   
     
<a href='#3'>3 Вывод</a> 

<a id='1'></a>
## Изучение данных

In [1]:
import pandas as pd
from sqlalchemy import create_engine

db_config = {'user': 'praktikum_student', 
'pwd': 'Sdf4$2;d-d30pp', 
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, 
'db': 'data-analyst-final-project-db'} 
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 
pd.set_option('display.max_colwidth', None)

<a id='1.1'></a>
### Изучение Таблицы books

In [2]:
query = ''' 
        SELECT *
        FROM books
        
        '''

pd.io.sql.read_sql(query, con = engine) 

book_id  author_id  \
0          1        546   
1          2        465   
2          3        407   
3          4         82   
4          5        125   
..       ...        ...   
995      996        571   
996      997        454   
997      998        201   
998      999         94   
999     1000        509   

                                                                                title  \
0                                                                        'Salem's Lot   
1                                                  1 000 Places to See Before You Die   
2                                 13 Little Blue Envelopes (Little Blue Envelope  #1)   
3                               1491: New Revelations of the Americas Before Columbus   
4                                                                                1776   
..                                                                                ...   
995                                          Wyrd Sisters (Discworld  #6; Witches #2)   
996                                                       Xenocide (Ender's Saga  #3)   
997                                                                   Year of Wonders   
998                                                       You Suck (A Love Story  #2)   
999  Zen and the Art of Motorcycle Maintenance: An Inquiry Into Values (Phaedrus  #1)   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

<a id='1.2'></a>
### Изучение Таблицы authors

In [3]:
query = ''' 
        SELECT *
        FROM authors
        LIMIT 5
        '''

pd.io.sql.read_sql(query, con = engine) 

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

<a id='1.3'></a>
### Изучение Таблицы publishers

In [4]:
query = ''' 
        SELECT *
        FROM publishers
        LIMIT 5
        '''

pd.io.sql.read_sql(query, con = engine) 

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

<a id='1.4'></a>
### Изучение Таблицы ratings

In [5]:
query = ''' 
        SELECT *
        FROM ratings
        LIMIT 5
        '''

pd.io.sql.read_sql(query, con = engine) 

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

<a id='1.5'></a>
### Изучение Таблицы reviews

In [6]:
query = ''' 
        SELECT *
        FROM reviews
        LIMIT 5
        '''

pd.io.sql.read_sql(query, con = engine) 

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                                                                                               text  
0                                        Mention society tell send professor analysis. Over provide race technology continue these.  
1                                                 Foot glass pretty audience hit themselves. Among admit investment argue security.  
2  Listen treat keep worry. Miss husband tax but person sport treatment industry. Kitchen decision deep the. Social party body the.  
3          Finally month interesting blue could nature cultural bit. Prepare beat finish grow that smile teach. Dream me play near.  
4                  Nation purpose heavy give wait song will. List dinner another whole positive radio fast. Music staff many green.

<a id='2'></a>
## Задания

<a id='2.1'></a>
### Сколько книг вышло после 1 января 2000 года

In [7]:
query = ''' 
        SELECT count (DISTINCT book_id)
        FROM books
        WHERE publication_date >= '2000-01-02'
        '''

pd.io.sql.read_sql(query, con = engine) 

count
0    819

<a id='2.2'></a>
### Для каждой книги посчитать количество обзоров и среднюю оценку

In [10]:
query = ''' 
        SELECT b.title,
               b.book_id,
               count (DISTINCT review_id) AS review,
               avg (rating) AS rating 
        FROM books AS b
        LEFT JOIN reviews ON reviews.book_id = b.book_id
        LEFT JOIN ratings ON ratings.book_id  = b.book_id
        GROUP BY b.book_id
        ORDER BY COUNT(reviews) DESC
        '''
pd.io.sql.read_sql(query, con = engine) 

title  book_id  \
0                                        Twilight (Twilight  #1)      948   
1                            The Hobbit  or There and Back Again      750   
2                                         The Catcher in the Rye      673   
3    Harry Potter and the Prisoner of Azkaban (Harry Potter  #3)      302   
4     Harry Potter and the Chamber of Secrets (Harry Potter  #2)      299   
..                                                           ...      ...   
995                                    Essential Tales and Poems      221   
996                                         Leonardo's Notebooks      387   
997             The Cat in the Hat and Other Dr. Seuss Favorites      672   
998              Anne Rice's The Vampire Lestat: A Graphic Novel       83   
999                                      The Natural Way to Draw      808   

     review    rating  
0         7  3.662500  
1         6  4.125000  
2         6  3.825581  
3         6  4.414634  
4         6  4.287500  
..      ...       ...  
995       0  4.000000  
996       0  4.000000  
997       0  5.000000  
998       0  3.666667  
999       0  3.000000  

[1000 rows x 4 columns]

<a id='2.3'></a>
### Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [11]:
query = ''' 
        SELECT p.publisher,
               count (DISTINCT book_id) AS books
        FROM publishers AS p
        LEFT JOIN books ON books.publisher_id = p.publisher_id
        WHERE num_pages > 50
        GROUP BY p.publisher
        ORDER BY books DESC
        LIMIT 2
        '''
pd.io.sql.read_sql(query, con = engine) 

publisher  books
0  Penguin Books     42
1        Vintage     31

<a id='2.4'></a>
### Определить автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками

In [12]:
query = ''' 
        WITH 
        i as (SELECT author,
                     b.book_id,
                     AVG(rating) AS rating
              FROM authors AS a
              LEFT JOIN books AS b ON b.author_id = a.author_id
              LEFT JOIN ratings AS r ON r.book_id = b.book_id
              GROUP BY author,
                       b.book_id
              HAVING COUNT(rating) > 50)
        SELECT i.author,
               AVG(rating) as rating
        FROM i
        GROUP BY author
        ORDER BY rating DESC
        LIMIT 2
               
        ''' 

pd.io.sql.read_sql(query, con = engine) 

author    rating
0         J.K. Rowling/Mary GrandPré  4.283844
1  Markus Zusak/Cao Xuân Việt Khương  4.264151

<a id='2.5'></a>
### Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [13]:
query = ''' 
       WITH 
       i as (SELECT count (review_id) as review
             FROM reviews
             WHERE username IN (
                                SELECT username
                                FROM ratings
                                GROUP BY username
                                HAVING COUNT(rating_id) > 50)
              GROUP BY username)
SELECT AVG(review) as avg_review
FROM i

'''
pd.io.sql.read_sql(query, con = engine) 

avg_review
0   24.333333

<a id='3'></a>
## Вывод

1. С 1 января 2000 года вышло 819 книг.
2. Наибольшее количество обзоров приходится на книгу Twilight (Twilight #1) - их 7, средняя оценка книги 3.66.
3. Penguin Books - издательство, которое выпустило самое большое количество книг с количеством странци более 50.
4. J.K. Rowling/Mary GrandPré самый популярный автор ( плюс художница к книге) - рейтин 4,28.
5. В среднем пользователи, которые ставили больше 50 оценок, публикуют 24 отзыва.